src/k8s/pkg/k8sd/api/cluster_join.go

```diff
@@ -26,6 +26,21 @@ func (e *Endpoints) postClusterJoin(s *state.State, r *http.Request) response.Re
 	internalToken := types.InternalWorkerNodeToken{}
 	// Check if token is worker token
 	if internalToken.Decode(req.Token) == nil {
+		// Check Server Auth
+		// Get remote certificate from the cluster member
+		cert, err := utils.GetRemoteCertificate(req.Address)
+		if err != nil {
+			return response.InternalError(fmt.Errorf("failed to get certificate of cluster member: %w", err))
+		}
+
+		// verify that the fingerprint of the certificate matches the fingerprint of the token
+		fingerprint := utils.CertFingerprint(cert)
+		if fingerprint != internalToken.Fingerprint {
```

Comment:
This should not happen here, but in https://github.com/canonical/k8s-snap/blob/[main](https://github.com/canonical/k8s-snap/blob/b7ae955ee3551d72119d9a46128e09cc5870ebdf/src/k8s/pkg/k8sd/app/hooks_bootstrap.go#L65)/src/k8s/pkg/k8sd/app/hooks_bootstrap.go#L65

so that we can also then create a client with the retrieved certificate